lib/data_schema.ex

```diff
@@ -918,10 +979,59 @@ defmodule DataSchema do
     Map.put(map, field, item)
   end
 
-  # This just lets us use either a module name for the data type OR a one arity fn.
-  defp call_cast_fn({module, fun, args}, value), do: apply(module, fun, [value | args])
-  defp call_cast_fn(module, value) when is_atom(module), do: module.cast(value)
-  defp call_cast_fn(fun, value) when is_function(fun, 1), do: fun.(value)
+  defp call_cast_fn({module, fun, args}, value, original_field) do
+    try do
+      apply(module, fun, [value | args])
+    rescue
+      error ->
+        message = error_message(original_field, error)
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
+
+  defp call_cast_fn(module, value, original_field) when is_atom(module) do
+    try do
+      module.cast(value)
+    rescue
+      error ->
+        message = error_message(original_field, error)
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
+
+  defp call_cast_fn(fun, value, original_field) when is_function(fun, 1) do
+    try do
+      fun.(value)
+    rescue
+      error ->
+        message = error_message(original_field, error)
+
+        reraise DataSchema.CastFunctionError, [message: message], __STACKTRACE__
+    end
+  end
+
+  # It would be much nicer if we could print out exactly what the original field looked
+  # like in the source code butwe have to fallback to inspect - meaning functions would
+  # look a bit shit. Ah well.
+  defp error_message({:field, original_field}, error) do
+    error_struct = error.__struct__
+      """
+      \n  ** (#{inspect(error_struct)})
```

Comment:
Once decided this should never change in case users of the library want to match on the specific error we are wrapping.
It's still a bit brittle